src/GraphQL/Validation/Rules/UniqueArgumentNames.cs

```diff
@@ -20,28 +20,24 @@ public class UniqueArgumentNames : IValidationRule
 
         /// <inheritdoc/>
         /// <exception cref="UniqueArgumentNamesError"/>
-        public Task<INodeVisitor> ValidateAsync(ValidationContext context)
-        {
-            var knownArgs = new Dictionary<string, Argument>();
+        public Task<INodeVisitor> ValidateAsync(ValidationContext context) => _nodeVisitor;
 
-            return new EnterLeaveListener(_ =>
+        private static readonly Task<INodeVisitor> _nodeVisitor = new NodeVisitors(
+            new MatchingNodeVisitor<Field>((__, context) => context.TypeInfo.UniqueArgumentNames_KnownArgs?.Clear()),
```

Comment:
It is necessary to write in the documentation that validation rules (at least standard ones) cannot be run in parallel.

Reply:
Why?  With the current design, it's impossible to run them in parallel if you wanted to.

Reply:
Someone could write their own validator.

Reply:
Well, it's not the validation rules that are a problem.  It's `TypeInfo`, returned as part of `ValidationContext`.  The validation rules are bits of static code returned as a static property.  Anything declared static is assumed to be mulithreading-safe, and so it is.

But I really don't know what scenario you could foresee anyone writing a validator that runs in parallel.  All of the node visitors rely on `TypeInfo`.  `TypeInfo` contains state information about the current node being walked.  If there was any type of parallel processing, either each walker needs its own `TypeInfo` (which would work perfectly fine and all the validation rules would run asynchronously), or theoretically the validation rules could run "asynchronously" within each node.  But this would require a separate thread for each one (starting each one with `Task.Run`), which would kill performance.  Nobody smart enough to write their own validator is stupid enough to put `Task.Run` in a tight loop.

On top of all of that, it is a given standard in .NET that all classes are not designed to be multi-threaded, unless explicitly stated otherwise.  `TypeInfo` is no different.  If anything, we should write comments that says which classes ARE designed for multithreaded use.

I'm not writing any comments to this effect.  Feel free to do so if you think it's of value; I have no argument there against additional documentation.  But I really don't know what I'd write if I tried to write something.

Reply:
```
    /// <summary>
    /// Provides information pertaining to the current state of the AST tree while being walked.
    /// Thus, validation rules checking is designed for sequential execution.
    /// </summary>
    public class TypeInfo : INodeVisitor
```

will be enougth

Reply:
Looks good. Not at a pc now. Feel free to add it now or I can later.

Reply:
done